File: _plugins/platforms.rb
# Shared access to the platform taxonomy defined in _config.yml.
#
# A platform is a top-level content directory (`<Platform>/<Product>/index.html`),
# so this list doubles as "which directories hold galleries". Every generator,
# template and script reads it from one place; see the `platforms:` key in
# _config.yml.
#
# Missing or malformed config raises rather than defaulting. An empty platform
# list would build a structurally valid but completely empty site, which is a
# far worse failure than a loud one.

module Jekyll
  module Platforms
    class ConfigError < StandardError; end

    def self.for(site)
      raw = site.config['platforms']

      unless raw.is_a?(Array) && !raw.empty?
        raise ConfigError, "_config.yml `platforms:` must be a non-empty list of " \
                           "directory names (got #{raw.inspect})"
      end

      raw.map do |name|
        unless name.is_a?(String) && !name.strip.empty?
          raise ConfigError, "_config.yml `platforms:` entries must be non-empty " \
                             "strings (got #{name.inspect})"
        end
        name
      end
    end

    # The platform a repo-relative path belongs to, or nil if it isn't inside one.
    #
    # Matches on the first path segment, never a substring: product folders like
    # "Universal Studios" or "T-Mobile" contain platform names as substrings, and
    # substring matching would file them under the wrong platform.
    def self.from_path(site, path)
      segments = path.to_s.sub(%r{\A/}, '').split('/')
      return nil if segments.empty?

      root = segments.first
      self.for(site).include?(root) ? root : nil
    end
  end
end

File: scripts/lib/platforms.mjs
/**
 * platforms.mjs
 *
 * Shared access to the platform taxonomy defined in _config.yml, so the Node
 * tooling reads the same list as the Jekyll build (see _plugins/platforms.rb).
 *
 * A platform is a top-level content directory — `<Platform>/<Product>/index.html`
 * — so this list also answers "which directories hold galleries".
 *
 * Missing or malformed config throws rather than defaulting: an empty list would
 * make every script silently find zero galleries and report success.
 */

import fs from "node:fs";
import path from "node:path";
import yaml from "js-yaml";

export function repoRoot() {
  return path.resolve(path.dirname(new URL(import.meta.url).pathname), "..", "..");
}

let cached = null;
let cachedConfig = null;

function readConfig(root) {
  if (cachedConfig) return cachedConfig;
  const configPath = path.join(root, "_config.yml");
  try {
    cachedConfig = yaml.load(fs.readFileSync(configPath, "utf-8")) || {};
  } catch (err) {
    throw new Error(`Could not read ${configPath}: ${err.message}`);
  }
  return cachedConfig;
}

export function loadPlatforms(root = repoRoot()) {
  if (cached) return cached;

  const platforms = readConfig(root).platforms;
  if (!Array.isArray(platforms) || platforms.length === 0) {
    throw new Error(
      `_config.yml \`platforms:\` must be a non-empty list of directory names ` +
        `(got ${JSON.stringify(platforms)})`
    );
  }
  for (const name of platforms) {
    if (typeof name !== "string" || name.trim() === "") {
      throw new Error(
        `_config.yml \`platforms:\` entries must be non-empty strings ` +
          `(got ${JSON.stringify(name)})`
      );
    }
  }

  cached = platforms;
  return cached;
}

/**
 * Platforms whose screenshots are phone-shaped, from `phone_platforms:`.
 *
 * Every entry must also appear in `platforms:`. Defaults to empty rather than
 * throwing: a site with no phone platforms is legitimate.
 */
export function loadPhonePlatforms(root = repoRoot()) {
  const raw = readConfig(root).phone_platforms;
  if (raw == null) return [];
  if (!Array.isArray(raw)) {
    throw new Error(
      `_config.yml \`phone_platforms:\` must be a list (got ${JSON.stringify(raw)})`
    );
  }
  const known = loadPlatforms(root);
  for (const name of raw) {
    if (!known.includes(name)) {
      throw new Error(
        `_config.yml \`phone_platforms:\` lists ${JSON.stringify(name)}, which is ` +
          `not in \`platforms:\` (${known.join(", ")})`
      );
    }
  }
  return raw;
}

/**
 * The platform a repo-relative path belongs to, or null if it isn't inside one.
 *
 * Matches the first path segment, never a substring — product folders such as
 * "T-Mobile" or "Universal Studios" contain platform names as substrings.
 */
export function platformFromPath(relPath, root = repoRoot()) {
  const first = relPath.split(path.sep).join("/").replace(/^\//, "").split("/")[0];
  return loadPlatforms(root).includes(first) ? first : null;
}
